stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/http_cache.bal

```diff
@@ -0,0 +1,131 @@
+// under the License.
+
+package ballerina.net.http;
+
+import ballerina.caching;
+
+struct HttpCache {
+    caching:Cache cache;
+    boolean isShared;
+}
+
+function createHttpCache (string name, int expiryTimeMillis, int capacity, float evictionFactor, boolean isShared) (HttpCache) {
+    HttpCache httpCache = {};
+    caching:Cache backingCache = caching:createCache(name, expiryTimeMillis, capacity, evictionFactor);
+    httpCache.cache = backingCache;
+    httpCache.isShared = isShared;
+    return httpCache;
+}
+
+function <HttpCache httpCache> put (string key, RequestCacheControl requestCacheControl, InResponse inboundResponse) {
+    if (inboundResponse.cacheControl.noStore ||
+        requestCacheControl.noStore ||
+        inboundResponse.cacheControl.isPrivate && httpCache.isShared) {
+        // TODO: Need to consider https://tools.ietf.org/html/rfc7234#section-3.2 as well here
+        return;
+    }
+
+    // Based on https://tools.ietf.org/html/rfc7234#page-6
+    // TODO: Consider cache control extensions as well here
+    if (inboundResponse.getHeader(EXPIRES) != null ||
+        inboundResponse.cacheControl.maxAge >= 0 ||
+        (inboundResponse.cacheControl.sMaxAge >= 0 && httpCache.isShared) ||
+        isCacheableStatusCode(inboundResponse.statusCode) ||
+        !inboundResponse.cacheControl.isPrivate) {
+
+        addEntry(httpCache.cache, key, inboundResponse);
+    }
+}
+
+function <HttpCache httpCache> get (string key) (InResponse) {
+    var cacheEntry, _ = (InResponse[])httpCache.cache.get(key);
+    return (cacheEntry == null) ? null : cacheEntry[lengthof cacheEntry - 1];
+}
+
+function <HttpCache httpCache> getAll (string key) (InResponse[]) {
+    var cacheEntry, _ = (InResponse[])httpCache.cache.get(key);
+    return cacheEntry;
+}
+
+function <HttpCache httpCache> getAllByETag (string key, string etag) (InResponse[]) {
+    InResponse[] cachedResponses = httpCache.getAll(key);
+    InResponse[] matchingResponses = [];
+    int i = 0;
+
+    foreach cachedResp in cachedResponses {
+        if (cachedResp.getHeader(ETAG) == etag && !etag.hasPrefix(WEAK_VALIDATOR_TAG)) {
+            matchingResponses[i] = cachedResp;
+            i = i + 1;
+        }
+    }
+
+    return matchingResponses;
+}
+
+function <HttpCache httpCache> getAllByWeakETag (string key, string etag) (InResponse[]) {
+    InResponse[] cachedResponses = httpCache.getAll(key);
+    InResponse[] matchingResponses = [];
+    int i = 0;
+
+    foreach cachedResp in cachedResponses {
+        if (weakValidatorEquals(etag, cachedResp.getHeader(ETAG))) {
+            matchingResponses[i] = cachedResp;
+            i = i + 1;
+        }
+    }
+
+    return matchingResponses;
+}
+
+function <HttpCache httpCache> remove(string key) {
+    httpCache.cache.remove(key);
+}
+
+function isCacheableStatusCode (int statusCode) (boolean) {
+    return statusCode == RESPONSE_200_OK || statusCode == RESPONSE_203_NON_AUTHORITATIVE_INFORMATION ||
```

Comment:
Shall we introduce index array for cachable status codes Instead of checking against the each of cacheable code?
Assume you have a boolean index array with cachable status code then you can simplify the logic as below.

return cachableStatusCodeIndex[statusCode] == true;

Reply:
This will result in a sparse array though, plus there are only the status codes mentioned here. Hence the individual checks.